src/box/memtx_tx.c

```diff
@@ -1740,27 +1751,68 @@ memtx_tx_story_insert_is_visible(struct memtx_story *story, struct txn *txn,
 }
 
 /**
- * Check whether the end of a @a story (that is deletion of its tuple) is
- * visible for transaction @a txn.
- * @param is_prepared_ok - whether prepared, not confirmed change is acceptable.
- * @param own_change - return true if the change was made by @txn itself.
- * @return true if the story end is visible, false otherwise.
+ * Check whether a @a story is deleted by a transaction @a txn.
+ * @param is_own_change - return true if the change was made by @txn itself,
+ * and this is guaranteed to always remain true (due to the presence of a
+ * read tracker or some other conditions).
+ * @return true if a story is deleted by a transaction, false otherwise.
  */
 static bool
-memtx_tx_story_delete_is_visible(struct memtx_story *story, struct txn *txn,
-				 bool is_prepared_ok, bool *is_own_change)
+memtx_tx_story_is_deleted_by_txn(struct memtx_story *story, struct txn *txn,
+				 bool *is_own_change)
 {
-	*is_own_change = false;
-
 	struct txn_stmt *dels = story->del_stmt;
 	while (dels != NULL) {
 		if (dels->txn == txn) {
-			/* Tuple is deleted by us (@txn). */
-			*is_own_change = true;
+			if (story->add_stmt != NULL &&
+			    story->add_stmt->txn == txn) {
+				*is_own_change = true;
+				return true;
+			}
+			/*
+			 * NOTE: If in some future MVCC update we can guarantee
+			 * that for a given statement stmt of transaction @txn,
+			 * stmt->del_story will always point to @story by means
+			 * other than:
+			 *  - story->add_stmt being a statement of transaction
+			 * @txn
+			 *  - story was deleted by a DELETE statement or a
+			 * REPLACE with an on_replace trigger set (a tracker was
+			 * created)
+			 * , then this code will need to be updated.
+			 */
```

Comment:
I don't think we need this comment here. Comment imagining what MVCC could be right in the middle of a function seems very distracting.

It could be nice just to write here that we rely on read trackers of own tuples.